File: src/net/minecraft/client/render/entity/CreeperEntityRenderer.cpp
#include "net/minecraft/client/render/entity/EntityRenderers.hpp"
#include "net/minecraft/client/gl/GlState.hpp"
#include "net/minecraft/client/render/entity/model/CreeperEntityModel.hpp"
#include "net/minecraft/entity/mob/CreeperEntity.hpp"
#include "net/minecraft/util/math/MathHelper.hpp"
#include <algorithm>
namespace net::minecraft::client::render::entity {
CreeperEntityRenderer::CreeperEntityRenderer() : LivingEntityRenderer(new model::CreeperEntityModel(), 0.5f) {
  chargedModel_ = new model::CreeperEntityModel(2.0f);
}
CreeperEntityRenderer::~CreeperEntityRenderer() {
  delete chargedModel_;
  chargedModel_ = nullptr;
}
void CreeperEntityRenderer::applyScale(const net::minecraft::LivingEntity& entity, float tickDelta) {
  const auto* creeper = dynamic_cast<const net::minecraft::entity::mob::CreeperEntity*>(&entity);
  float swell = creeper != nullptr ? creeper->getScale(tickDelta) : 0.0f;
  float pulse = 1.0f + net::minecraft::util::math::MathHelper::sin(swell * 100.0f) * swell * 0.01f;
  if(swell < 0.0f)
    swell = 0.0f;
  if(swell > 1.0f)
    swell = 1.0f;
  swell *= swell;
  swell *= swell;
  const float sx = (1.0f + swell * 0.4f) * pulse;
  const float sy = (1.0f + swell * 0.1f) / pulse;
  gl::scalef(sx, sy, sx);
}
int CreeperEntityRenderer::getOverlayColor(const net::minecraft::LivingEntity& entity, float /*brightness*/,
                                           float tickDelta) const {
  const auto* creeper = dynamic_cast<const net::minecraft::entity::mob::CreeperEntity*>(&entity);
  const float swell = creeper != nullptr ? creeper->getScale(tickDelta) : 0.0f;
  if(static_cast<int>(swell * 10.0f) % 2 == 0)
    return 0;
  int alpha = static_cast<int>(swell * 0.2f * 255.0f);
  alpha = std::clamp(alpha, 0, 255);
  return (alpha << 24) | (255 << 16) | (255 << 8) | 255;
}
bool CreeperEntityRenderer::bindTexture(const net::minecraft::LivingEntity& entity, int layer, float tickDelta) {
  const auto* creeper = dynamic_cast<const net::minecraft::entity::mob::CreeperEntity*>(&entity);
  if(creeper == nullptr || !creeper->isCharged())
    return false;
  if(layer == 1) {
    const float time = static_cast<float>(entity.age) + tickDelta;
    EntityRenderer::bindTexture("/armor/power.png");
    gl::matrixMode(5890);
    gl::loadIdentity();
    const float offset = time * 0.01f;
    gl::translatef(offset, offset, 0.0f);
    setDecorationModel(chargedModel_);
    gl::matrixMode(5888);
    gl::setCap(gl::cap::Blend, true);
    gl::color4f(0.5f, 0.5f, 0.5f, 1.0f);
    gl::blendFunc(gl::blend::One, gl::blend::One);
    return true;
  }
  if(layer == 2) {
    gl::matrixMode(5890);
    gl::loadIdentity();
    gl::matrixMode(5888);
    gl::setCap(gl::cap::Blend, false);
  }
  return false;
}
bool CreeperEntityRenderer::bindDecorationTexture(const net::minecraft::LivingEntity& /*entity*/, int /*layer*/,
                                                  float /*tickDelta*/) {
  return false;
}
} // namespace net::minecraft::client::render::entity
#include "net/minecraft/client/entity/EntityClientRendererRegistration.hpp"
#include "net/minecraft/entity/mob/CreeperEntity.hpp"
namespace net::minecraft::entity::mob {
std::unique_ptr<::net::minecraft::client::render::entity::EntityRenderer> CreeperEntity::ClientRenderer::create() {
  return std::make_unique<::net::minecraft::client::render::entity::CreeperEntityRenderer>();
}
} // namespace net::minecraft::entity::mob
namespace {
static ::net::minecraft::registry::RegisterEntityRenderer<net::minecraft::entity::mob::CreeperEntity> autoRendererReg;
} // namespace

File: mcp/src/net/minecraft/client/render/entity/CreeperEntityRenderer.java
/*
 * Decompiled with CFR 0.1.1 (FabricMC 57d88659).
 * 
 * Could not load the following classes:
 *  net.fabricmc.api.EnvType
 *  net.fabricmc.api.Environment
 */
package net.minecraft.client.render.entity;

import net.fabricmc.api.EnvType;
import net.fabricmc.api.Environment;
import net.minecraft.client.render.entity.LivingEntityRenderer;
import net.minecraft.client.render.entity.model.CreeperEntityModel;
import net.minecraft.client.render.entity.model.EntityModel;
import net.minecraft.entity.mob.CreeperEntity;
import net.minecraft.util.math.MathHelper;
import org.lwjgl.opengl.GL11;

@Environment(value=EnvType.CLIENT)
public class CreeperEntityRenderer
extends LivingEntityRenderer {
    private EntityModel model = new CreeperEntityModel(2.0f);

    public CreeperEntityRenderer() {
        super(new CreeperEntityModel(), 0.5f);
    }

    protected void applyScale(CreeperEntity arg, float f) {
        CreeperEntity creeperEntity = arg;
        float f2 = creeperEntity.getScale(f);
        float f3 = 1.0f + MathHelper.sin(f2 * 100.0f) * f2 * 0.01f;
        if (f2 < 0.0f) {
            f2 = 0.0f;
        }
        if (f2 > 1.0f) {
            f2 = 1.0f;
        }
        f2 *= f2;
        f2 *= f2;
        float f4 = (1.0f + f2 * 0.4f) * f3;
        float f5 = (1.0f + f2 * 0.1f) / f3;
        GL11.glScalef(f4, f5, f4);
    }

    protected int getOverlayColor(CreeperEntity arg, float f, float g) {
        CreeperEntity creeperEntity = arg;
        float f2 = creeperEntity.getScale(g);
        if ((int)(f2 * 10.0f) % 2 == 0) {
            return 0;
        }
        int n = (int)(f2 * 0.2f * 255.0f);
        if (n < 0) {
            n = 0;
        }
        if (n > 255) {
            n = 255;
        }
        int n2 = 255;
        int n3 = 255;
        int n4 = 255;
        return n << 24 | n2 << 16 | n3 << 8 | n4;
    }

    protected boolean bindTexture(CreeperEntity arg, int i, float f) {
        if (arg.isCharged()) {
            if (i == 1) {
                float f2 = (float)arg.age + f;
                this.bindTexture("/armor/power.png");
                GL11.glMatrixMode(5890);
                GL11.glLoadIdentity();
                float f3 = f2 * 0.01f;
                float f4 = f2 * 0.01f;
                GL11.glTranslatef(f3, f4, 0.0f);
                this.setDecorationModel(this.model);
                GL11.glMatrixMode(5888);
                GL11.glEnable(3042);
                float f5 = 0.5f;
                GL11.glColor4f(f5, f5, f5, 1.0f);
                GL11.glDisable(2896);
                GL11.glBlendFunc(1, 1);
                return true;
            }
            if (i == 2) {
                GL11.glMatrixMode(5890);
                GL11.glLoadIdentity();
                GL11.glMatrixMode(5888);
                GL11.glEnable(2896);
                GL11.glDisable(3042);
            }
        }
        return false;
    }

    protected boolean bindDecorationTexture(CreeperEntity arg, int i, float f) {
        return false;
    }
}

